lib/active_model/serializer/adapter/json_api/link.rb

```diff
@@ -10,29 +10,40 @@ def initialize(serializer)
 
           def href(value)
             self._href = value
+            nil
           end
 
           def meta(value)
             self._meta = value
+            nil
           end
 
-          def string(value)
-            self._string = value
+          def value(value)
+            if value.respond_to?(:call)
+              string instance_eval(&value)
+              _string || to_hash
```

Comment:
Oh actually, your idea of using the return value of the block makes for quite a cool DSL. My suggestion of moving the actual computing of the value inside the initializer still stands, but other than this I think you're right, this is the way to go.